File: src/solutions/2023/10/go/main.go
package main

import (
	core "advent/of/go/src/langs/go"
	"os"
	"strconv"
	"strings"
)

func main() {
	options := core.SolutionOptions{
		HasAlternate: true,
		HasIo:        false,
	}
	core.Run(os.Args, part1, part2, options)
}

func findStart(grid []string) [2]int64 {
	for y := 0; y < len(grid); y++ {
		for x := 0; x < len(grid[y]); x++ {
			if grid[y][x] == 'S' {
				return [2]int64{int64(x), int64(y)}
			}
		}
	}
	panic("Couldn't find starting point")
}

func goUp(
	grid []string,
	x int64,
	y int64,
) bool {
	criteria := "S7F|"
	return y > 0 && strings.Contains(criteria, string(grid[y-1][x]))
}

func goDown(
	grid []string,
	x int64,
	y int64,
) bool {
	criteria := "SLJ|"
	return y < int64(len(grid))-1 && strings.Contains(criteria, string(grid[y+1][x]))
}

func goLeft(
	grid []string,
	x int64,
	y int64,
) bool {
	criteria := "SLF-"
	return x > 0 && strings.Contains(criteria, string(grid[y][x-1]))
}

func goRight(
	grid []string,
	x int64,
	y int64,
) bool {
	criteria := "S7J-"
	return x < int64(len(grid[y]))-1 && strings.Contains(criteria, string(grid[y][x+1]))
}

func lookUp(
	grid []string,
	x int64,
	y int64,
) [][2]int64 {
	char := grid[y][x]
	res := make([][2]int64, 0)
	switch char {
	case '|':
		if goUp(grid, x, y) {
			res = append(res, [2]int64{x, y - 1})
		}
		if goDown(grid, x, y) {
			res = append(res, [2]int64{x, y + 1})
		}
	case '-':
		if goLeft(grid, x, y) {
			res = append(res, [2]int64{x - 1, y})
		}
		if goRight(grid, x, y) {
			res = append(res, [2]int64{x + 1, y})
		}
	case 'L':
		if goUp(grid, x, y) {
			res = append(res, [2]int64{x, y - 1})
		}
		if goRight(grid, x, y) {
			res = append(res, [2]int64{x + 1, y})
		}
	case 'J':
		if goUp(grid, x, y) {
			res = append(res, [2]int64{x, y - 1})
		}
		if goLeft(grid, x, y) {
			res = append(res, [2]int64{x - 1, y})
		}
	case '7':
		if goDown(grid, x, y) {
			res = append(res, [2]int64{x, y + 1})
		}
		if goLeft(grid, x, y) {
			res = append(res, [2]int64{x - 1, y})
		}
	case 'F':
		if goDown(grid, x, y) {
			res = append(res, [2]int64{x, y + 1})
		}
		if goRight(grid, x, y) {
			res = append(res, [2]int64{x + 1, y})
		}
	case 'S':
		if goUp(grid, x, y) {
			res = append(res, [2]int64{x, y - 1})
		}
		if goDown(grid, x, y) {
			res = append(res, [2]int64{x, y + 1})
		}
		if goLeft(grid, x, y) {
			res = append(res, [2]int64{x - 1, y})
		}
		if goRight(grid, x, y) {
			res = append(res, [2]int64{x + 1, y})
		}
	}
	return res
}

func part1(input string, isTest bool) string {
	grid := strings.Split(input, "\n")
	visited := make([][]bool, len(grid))
	for i := 0; i < len(grid); i++ {
		visited[i] = make([]bool, len(grid[i]))
		for j := 0; j < len(grid[i]); j++ {
			visited[i][j] = false
		}
	}

	var i int64 = 0
	queue := [][2]int64{findStart(grid)}
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		found := lookUp(grid, current[0], current[1])
		for _, f := range found {
			if visited[f[1]][f[0]] {
				continue
			}
			visited[f[1]][f[0]] = true
			queue = append(queue, f)
			i++
		}
	}

	return strconv.FormatInt(i/2, 10)
}

func part2(input string, isTest bool) string {
	grid := strings.Split(input, "\n")
	visited := make([][]bool, len(grid)*3)
	for i := range visited {
		visited[i] = make([]bool, len(grid[0])*3)
		for j := range visited[i] {
			visited[i][j] = false
		}
	}

	queue := [][2]int64{findStart(grid)}
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		found := lookUp(grid, current[0], current[1])
		for _, f := range found {
			visited[f[1]*3+1+current[1]-f[1]][f[0]*3+1+current[0]-f[0]] = true
			if visited[f[1]*3+1][f[0]*3+1] {
				continue
			}
			visited[f[1]*3+1][f[0]*3+1] = true
			queue = append(queue, f)
		}
	}

	queue = append(queue, [2]int64{0, 0})
	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]
		if visited[current[1]][current[0]] {
			continue
		}
		visited[current[1]][current[0]] = true
		for _, ud := range [3]int64{-1, 0, 1} {
			for _, lr := range [3]int64{-1, 0, 1} {
				if current[1]+ud < 0 || current[1]+ud >= int64(len(visited)) {
					continue
				}
				if current[0]+lr < 0 || current[0]+lr >= int64(len(visited[0])) {
					continue
				}
				if visited[current[1]+ud][current[0]+lr] {
					continue
				}
				queue = append(queue, [2]int64{current[0] + lr, current[1] + ud})
			}
		}
	}

	// check result
	var res int64 = 0
	for y := range grid {
		for x := range grid[0] {
			if !visited[1+y*3][1+x*3] {
				res += 1
			}
		}
	}

	return strconv.FormatInt(res, 10)
}

File: src/solutions/2023/10/csharp/Main.cs

using Core;

namespace Year2023;
public static class Day10
{
   static SolutionOptions OPTIONS = new()
   {
      HasAlternate = true,
      HasIO = false,
   };

   private static (int x, int y) FindStart(string[] grid)
   {
      for (int y = 0; y < grid.Length; y++)
      {
         for (int x = 0; x < grid[y].Length; x++)
         {
            if (grid[y][x] == 'S')
            {
               return (x, y);
            }
         }
      }
      return (0, 0);
   }

   private static bool GoUp(
      string[] grid,
      int x,
      int y
   )
   {
      string criteria = "S7F|";
      return y > 0 && criteria.Contains(grid[y - 1][x]);
   }

   private static bool GoDown(
      string[] grid,
      int x,
      int y
   )
   {
      string criteria = "SLJ|";
      return y < grid.Length - 1 && criteria.Contains(grid[y + 1][x]);
   }

   private static bool GoLeft(
      string[] grid,
      int x,
      int y
   )
   {
      string criteria = "SLF-";
      return x > 0 && criteria.Contains(grid[y][x - 1]);
   }

   private static bool GoRight(
      string[] grid,
      int x,
      int y
   )
   {
      string criteria = "S7J-";
      return x < grid[y].Length - 1 && criteria.Contains(grid[y][x + 1]);
   }

   private static List<(int x, int y)> LookUp(
      string[] grid,
      int x,
      int y
   )
   {
      char c = grid[y][x];
      var res = new List<(int x, int y)>();
      if (c == '|')
      {
         if (GoUp(grid, x, y))
            res.Add((x, y - 1));
         if (GoDown(grid, x, y))
            res.Add((x, y + 1));
      }
      if (c == '-')
      {
         if (GoLeft(grid, x, y))
            res.Add((x - 1, y));
         if (GoRight(grid, x, y))
            res.Add((x + 1, y));
      }
      if (c == 'L')
      {
         if (GoUp(grid, x, y))
            res.Add((x, y - 1));
         if (GoRight(grid, x, y))
            res.Add((x + 1, y));
      }
      if (c == 'J')
      {
         if (GoUp(grid, x, y))
            res.Add((x, y - 1));
         if (GoLeft(grid, x, y))
            res.Add((x - 1, y));
      }
      if (c == '7')
      {
         if (GoDown(grid, x, y))
            res.Add((x, y + 1));
         if (GoLeft(grid, x, y))
            res.Add((x - 1, y));
      }
      if (c == 'F')
      {
         if (GoDown(grid, x, y))
            res.Add((x, y + 1));
         if (GoRight(grid, x, y))
            res.Add((x + 1, y));
      }
      if (c == 'S')
      {
         if (GoUp(grid, x, y))
            res.Add((x, y - 1));
         if (GoDown(grid, x, y))
            res.Add((x, y + 1));
         if (GoLeft(grid, x, y))
            res.Add((x - 1, y));
         if (GoRight(grid, x, y))
            res.Add((x + 1, y));
      }
      return res;
   }

   static string Part1(string input, bool isTest = false)
   {
      var grid = input.Split('\n');
      bool[][] visited = new bool[grid.Length][];
      for (int i = 0; i < grid.Length; i++)
      {
         visited[i] = new bool[grid[i].Length];
      }

      int res = 0;
      Queue<(int x, int y)> queue = new();
      queue.Enqueue(FindStart(grid));
      while (queue.Count > 0)
      {
         var (x, y) = queue.Dequeue();
         var found = LookUp(grid, x, y);
         foreach (var f in found)
         {
            if (visited[f.y][f.x])
               continue;
            visited[f.y][f.x] = true;
            queue.Enqueue(f);
            res++;
         }
      }

      return (res / 2).ToString();
   }

   static string Part2(string input, bool isTest = false)
   {
      var grid = input.Split('\n');
      bool[][] visited = new bool[grid.Length * 3][];
      for (int i = 0; i < visited.Length; i++)
      {
         visited[i] = new bool[grid[0].Length * 3];
      }

      Queue<(int x, int y)> queue = new();
      queue.Enqueue(FindStart(grid));
      while (queue.Count > 0)
      {
         var (x, y) = queue.Dequeue();
         var found = LookUp(grid, x, y);
         foreach (var f in found)
         {
            visited[f.y * 3 + 1 + y - f.y][
               f.x * 3 + 1 + x - f.x
            ] = true;
            if (visited[f.y * 3 + 1][f.x * 3 + 1]) continue;
            visited[f.y * 3 + 1][f.x * 3 + 1] = true;
            queue.Enqueue(f);
         }
      }

      queue.Enqueue((0, 0));
      while (queue.Count > 0)
      {
         var (x, y) = queue.Dequeue();
         if (visited[y][x]) continue;
         visited[y][x] = true;
         for (int ud = -1; ud <= 1; ud++)
         {
            for (int lr = -1; lr <= 1; lr++)
            {
               if (y + ud < 0 || y + ud >= visited.Length)
               {
                  continue;
               }
               if (x + lr < 0 || x + lr >= visited[0].Length)
               {
                  continue;
               }
               if (visited[y + ud][x + lr]) continue;
               queue.Enqueue((x + lr, y + ud));
            }
         }
      }

      int res = 0;
      for (int y = 0; y < grid.Length; y++)
      {
         for (int x = 0; x < grid[0].Length; x++)
         {
            if (!visited[1 + y * 3][1 + x * 3]) res++;
         }
      }

      return res.ToString();
   }

   static void Main(string[] args)
   {
      Runner.Run(args, Part1, Part2, OPTIONS);
   }
}
